express-notes-api/index.js

```diff
@@ -0,0 +1,199 @@
+    };
+    res.status(404).json(noId);
+
+  } else {
+    res.json(notesObj[targetNoteId]);
+  }
+});
+
+// Express.json parses incoming requests with JSON payloads
+appObj.use(express.json());
+
+// we need post a new note to the data.json file
+/*
+if the user doesn't put any content send an error message and status code
+if the content is not underfined then we need to post the note
+*/
+appObj.post('/api/notes', (req, res) => {
+  const newNote = req.body;
+
+  if (newNote.content === undefined) {
+    const errObj = {
+      error: 'content is a required field'
+    };
+    res.status(400).json(errObj);
+
+  } else if (newNote.content !== undefined) {
+    // set the new notes id to the current value of nextId from data.json
+    newNote.id = dataId;
+    // put the new note into the correct spot in the notes object on the data.json file
+    dataJSON.notes[dataId.toString()] = newNote;
+    // increment the the nextId from the data.json file to prepare for the
+    // note entry
+    dataJSON.nextId++;
+    // the new note is now part of data.json object and we need to
+    // put it into a json string to send it
+    const newNoteJSON = JSON.stringify(dataJSON, null, 2);
+
+    // we want to write the newNote in json string form to the data.json file
+    fs.writeFile('data.json', newNoteJSON, 'utf8', err => {
+
+      if (err) {
+        const postErr = {
+          error: 'an unexpected error occurred.'
+        };
+        console.error(postErr);
+        // if an error happens then the status 500 will be sent from the server
+        res.status(500).json(postErr);
+      }
+
+      // this will send the new note back to the client and the created status code
+      // we want to do this once the new note is written to data.json
+      res.status(201).send(newNote);
+    });
+  }
+});
+
+appObj.delete('/api/notes/:id', (req, res) => {
+  const noteId = req.params.id;
+
+  if (!Number.isInteger(parseInt(noteId)) || parseInt(noteId) <= 0) {
+    const errObj = {
+      error: 'id must be a positive integer'
+    };
+    res.status(400).json(errObj);
+
+  } else if (dataJSON.notes[noteId] === undefined) {
+    const noId = {
+      error: `cannot find note with id ${noteId}`
+    };
+    res.status(404).json(noId);
+
+  } else {
+    // we want to delete the note from the notes object at the client's
+    // desired id
+    delete dataJSON.notes[noteId];
+    // then we want to put the updated data.json object into a JSON string
+    const currentNotes = JSON.stringify(dataJSON, null, 2);
+
+    // now we write the updated data.json object to the data.json file
+    fs.writeFile('data.json', currentNotes, 'utf8', err => {
+
+      if (err) {
+        const postErr = {
+          error: 'an unexpected error occurred.'
+        };
+        console.error(postErr);
+        res.status(500).json(postErr);
+      }
+
+      // this will send the status to the client from the server
+      // no body will be shown
+      res.sendStatus(204);
```

Comment:
In each of your `fs.writeFile` callbacks, be sure to include either a `return` (guard) or an `else` block to guarantee that it's impossible for the server to attempt responding twice to the same request.